common/src/main/java/net/opentsdb/data/TimeSeries.java

```diff
@@ -0,0 +1,85 @@
+// This file is part of OpenTSDB.
+// Copyright (C) 2017  The OpenTSDB Authors.
+//
+// This program is free software: you can redistribute it and/or modify it
+// under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 2.1 of the License, or (at your
+// option) any later version.  This program is distributed in the hope that it
+// will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty
+// of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser
+// General Public License for more details.  You should have received a copy
+// of the GNU Lesser General Public License along with this program.  If not,
+// see <http://www.gnu.org/licenses/>.
+// This file is part of OpenTSDB.
+// Copyright (C) 2017  The OpenTSDB Authors.
+//
+// This program is free software: you can redistribute it and/or modify it
+// under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 2.1 of the License, or (at your
+// option) any later version.  This program is distributed in the hope that it
+// will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty
+// of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser
+// General Public License for more details.  You should have received a copy
+// of the GNU Lesser General Public License along with this program.  If not,
+// see <http://www.gnu.org/licenses/>.
+package net.opentsdb.data;
+
+import java.util.Collection;
+import java.util.Iterator;
+
+import com.google.common.reflect.TypeToken;
+
+/**
+ * A collection of data for a specific time series, i.e. {@link TimeSeriesId}.
+ * The collection may contain multiple data types, e.g. a series of numeric
+ * values and a series of strings or annotations.
+ * <p>
+ * To pull data out of this collection, call {@link #iterator(TypeToken)} or
+ * {@link #iterators()}. These will return Java iterators that return
+ * {@link TimeSeriesValue} objects on each call to {@link Iterator#next()}. Each
+ * call to one of these iterator methods will return a new view of the data 
+ * starting at the beginning of the set.
+ * <p>
+ * All data types will contain data within the same time range for this time
+ * series.
+ * 
+ * @since 3.0
+ */
+public interface TimeSeries {
+
+  /**
+   * A reference to the time series ID associated with this value.
+   * @return A non-null {@link TimeSeriesId}.
+   */
+  public TimeSeriesId id();
+  
+  /**
+   * Returns a new iterator for the type of time series data denoted by {code type}.
+   * @param type A non-null {@link TimeSeriesDataType} type of data.
+   * @return A new iterator for the type of data or null if no such type is
+   * present for the time series.
+   */
+  public Iterator<TimeSeriesValue<? extends TimeSeriesDataType>> 
+    iterator(final TypeToken<?> type);
```

Comment:
Can we return an Optional<Iterator<>>? That clarifies the intent via type.

Reply:
Good idea.